activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -146,7 +146,11 @@ def release_advisory_lock(lock_name) # :nodoc:
       end
 
       def native_database_types
-        NATIVE_DATABASE_TYPES
+        return NATIVE_DATABASE_TYPES unless supports_datetime_with_precision?
+
+        NATIVE_DATABASE_TYPES.dup.tap do |types|
+          type[:datetime][:precision] = 6
+        end
```

Comment:
This is just a code knit and not at all related to the feature you're trying to work on 😅 , but how about to write this as:

```
if supports_datetime_with_precision?
  return NATIVE_DATABASE_TYPES.dup.blahblah
end

return NATIVE_DATABASE_TYPES
```

The `return`s could be implicit I guess, but I try to avoid unless here if you're planning to add more types.

Reply:
I think it would be better to memoize this, same as we do on https://github.com/rails/rails/blob/main/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L345